### DISCUSSION TITLE
Move the documentation on some DeclarativeBase members

### DISCUSSION THREAD

**Discussion by lelit** [sqlalchemy/sqlalchemy#10480] [Ideas]
Currently, when using the `DeclarativeBase` in a project with its own Sphinx documentation extracted by `autodoc`, we get a warning like the following:
```
...
writing output... [100%] modules
... docstring of ...Base.metadata:6: WARNING: undefined label: 'orm_declarative_metadata'
generating indices... genindex py-modindex fatto
```

That is caused by the fact that the `Base` entity inherits from `DeclarativeBase`, and in some way `autodoc` _considers_ (I guess due to the fact that it uses the `DeclarativeAttributeIntercept` metaclass) `metadata` and `registry` as members of the `Base` class, and thus a copy of [metadata](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBase.metadata)'s [docstring](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/decl_api.py#L766). Of course, Sphinx in an external project _based_ on SA cannot resolve those references, and emits the warning above, producing a slightly ugly entries in the project's doc page for `Base`.

The simplest solution I can think of is moving the documentation of those members (`registry` and `metadata`) from the docstring to the [mapping_api.rst](https://github.com/sqlalchemy/sqlalchemy/blob/main/doc/build/orm/mapping_api.rst?plain=1#L112) file.

Does this sound reasonable?

**Comment by zzzeek** [comment]
per sphinx documentation, superclass attributes are not included when it autodocs a class:

https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

> By default, autodoc will not generate document for the members that are private, not having docstrings, inherited from super class, or special members.

if you are using `inherited-members`, then specify the bases you want to exclude:   "It can take an ancestor class not to document inherited members from it. By default, members of object class are not documented. To show them all, give None to the option":

```

.. autoclass: MyClass
    :inherited-members: DeclarativeBase

```

Overall we have other kinds of classes that are subclassable also like `UserDefinedType`, `TypeDecorator`, etc, and it's not reasonable that we have to remove 100% of inline documentation strings from all these classes to solve this kind of issue.   

if something special is happening because these attributes are from a metaclass of some kind that isn't usually happening, then we need to look at sphinx to see what's going on.

**Comment by lelit** [comment]
I'm not using `inherited-members`:

My `models/__init__.py` contains
```python
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    "Abstract base entity class."

    __abstract__ = True

    ...
```

and the corresponding Sphinx doc is:

```
==============================================
 :mod:`sol.models` -- SQLAlchemy modelization
==============================================

.. automodule:: sol.models

.. autoclass:: sol.models.Base
   :members:
```

As said, I think that Sphinx is tricked by the fact that `DeclarativeBase` has a metaclass that injects attributes in the class itself (but I may be very wrong on that).

Anyway, if you think this is not viable, I will try to come up with a different solution, maybe something like
```python
class AbstractBase(DeclarativeBase):
    pass

class Base(AbstractBase):
    "Abstract base entity class."

    __abstract__ = True

    ...
```

and omitting `AbstractBase` from my doc...

**Reply by zzzeek** [reply]
oh I see, you think it's the mechanics of the metaclass...let me check on that

**Reply by zzzeek** [reply]
these attributes are directly in DeclarativeBase in a TYPE_CHECKING block, which means that code isn't present at runtime.   then the metadata and registry attributes are assigned to the Base you create at runtime using the `__init_subclass__()` hook., but they don't have any docstrings at that level.    

so...I have no idea how Sphinx is even getting these docstrings unless they are setting TYPE_CHECKING at runtime, and...they shouldnt be here, because you are not auto-docing DeclarativeBase.    

we need to have these attributes declared in TYPE_CHECKING because they have type annotations.

it would be great if sphinx could clarify / improve what's happening here because it's a burden for us to have very particular documentation arrangements that have no way of being tested

**Reply by lelit** [reply]
Ok, indeed I could not figure out what was going on there, and assumed some little magic was happening, like some kind of "hidden mixin" with `DeclarativeBaseNoMeta`...
Given that this is trickier than expected and triggers my curiosity, I will do some experiment and either report back or open an issue with Sphinx.

**Comment by lelit** [comment]
I'm very surprised: I looked around, and then digged in [the source](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/importer.py#L105), and found nothing that enables `TYPE_CHECKING`.

I wrote a [simple gist](https://gist.github.com/lelit/6630bdbedfba31c8d8c214ea08e4ef8e), with the following Python module:
```python
import typing


class Foo:
    "This is a base class"

    metadata: dict = {}
    "This is something"

    if False:
        hidden: bool = True
        "This is something that should be visible only when TYPE_CHECKING is True"


class Bar(Foo):
    "This is a derivated class"

```

and guess what, the `hidden` attribute is there in the doc!

I'm going to open an issue/discussion on Sphinx, but apparently the inspection done by `autodoc` is a _plain traversal_ of the `AST` (just guessing here), and thus it does not honor the `if False:` condition.

**Reply by lelit** [reply]
A quick inspection of [pycode](https://github.com/sphinx-doc/sphinx/tree/master/sphinx/pycode), what `autodoc` uses to do its job, seems to confirm that: it does not care about _conditional blocks_.

**Reply by zzzeek** [reply]
That's probably the right behavior for them, because we like that sphinx does document docstrings that are in TYPE_CHECKING.    

I dont know that sphinx is going to have help for this, did you try adding DeclarativeBase as an ignore?  if you have that base as an ignore, I would think that should work.

**Reply by CaselIT** [reply]
It makes sense that the code is not executed while generating the documentation.
I wonder if there is a way of telling sphinx to ignore one particular attribute using some "magic" command in the doc string, something like `.. sphinx ignore`

**Reply by lelit** [reply]
I will probably go with `:exclude-members: metadata, registry` in my reST entry for the `Base` class.
Federico, I miss the point of your idea: even if there was such "magic", it would exclude the attribute in the SA docs too, wouldn't it?

Note also that the module code **is** executed (it is imported, see [the warning box at the top](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)), but the `autodoc` machinery still has to parse it again to extract _attributes_ docstrings. Maybe there could be a new option, say "ok-autoexclude-attributes-not-present-in-the-class-dir()" that intersects the _members_ collected by the `pycode` parser with `dir(imported_class)` or something like that? Would that makes sense? If so, I'll open a suggestion on the Sphinx repo...

**Reply by CaselIT** [reply]
> Federico, I miss the point of your idea: even if there was such "magic", it would exclude the attribute in the SA docs too, wouldn't it?

yes, but we could probably override them while generating the sqlalchemy docs. But it probably complicates stuff unnecessary 

> Note also that the module code is executed (it is imported, see [the warning box at the top](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)), but the autodoc machinery still has to parse it again to extract attributes docstrings.

I think the ast must be parsed in any case since there is no way of accessing the doc string of an attribute at runtime, since it's not kept by python

**Reply by lelit** [reply]
> I think the ast must be parsed in any case since there is no way of accessing the doc string of an attribute at runtime, since it's not kept by python

Right, that's what I meant.
